modules/vision/src/pose-estimation/vpPoseRGBD.cpp

```diff
@@ -0,0 +1,320 @@
+      if (W[i] < smallestSv) {
+        smallestSv = W[i];
+        indexSmallestSv = i;
+      }
+    }
+
+    normal = V.getCol(indexSmallestSv);
+
+    // Compute plane equation
+    double A = normal[0], B = normal[1], C = normal[2];
+    double D = -(A * centroid_x + B * centroid_y + C * centroid_z);
+
+    // Update plane equation
+    plane_equation_estimated[0] = A;
+    plane_equation_estimated[1] = B;
+    plane_equation_estimated[2] = C;
+    plane_equation_estimated[3] = D;
+
+    // Compute error points to estimated plane
+    prev_error = error;
+    error = 0.0;
+    for (unsigned int i = 0; i < nPoints; i++) {
+      residues[i] = std::fabs(A * point_cloud_face[3 * i] + B * point_cloud_face[3 * i + 1] +
+                    C * point_cloud_face[3 * i + 2] + D) / sqrt(A * A + B * B + C * C);
+      error += residues[i] * residues[i];
+    }
+    error /= sqrt(error / total_w);
+  }
+
+  // Update final weights
+  tukey.MEstimator(vpRobust::TUKEY, residues, weights);
+
+  // Update final centroid
+  centroid.resize(3, false);
+  double total_w = 0.0;
+
+  for (unsigned int i = 0; i < nPoints; i++) {
+    centroid[0] += weights[i] * point_cloud_face[3 * i];
+    centroid[1] += weights[i] * point_cloud_face[3 * i + 1];
+    centroid[2] += weights[i] * point_cloud_face[3 * i + 2];
+    total_w += weights[i];
+  }
+
+  centroid[0] /= total_w;
+  centroid[1] /= total_w;
+  centroid[2] /= total_w;
+
+  // Compute final plane equation
+  double A = normal[0], B = normal[1], C = normal[2];
+  double D = -(A * centroid[0] + B * centroid[1] + C * centroid[2]);
+
+  // Update final plane equation
+  plane_equation_estimated[0] = A;
+  plane_equation_estimated[1] = B;
+  plane_equation_estimated[2] = C;
+  plane_equation_estimated[3] = D;
+}
+
+double computeZMethod1(const vpColVector& plane_equation, double x, double y) {
+  return -plane_equation[3] / (plane_equation[0]*x + plane_equation[1]*y + plane_equation[2]);
+}
+
+bool validPose(const vpHomogeneousMatrix& cMo) {
+  bool valid = true;
+
+  for (unsigned int i = 0; i < cMo.getRows() && valid; i++) {
+    for (unsigned int j = 0; j < cMo.getCols() && valid; j++) {
+      if (vpMath::isNaN(cMo[i][j])) {
+        valid = false;
+      }
+    }
+  }
+
+  return valid;
+}
+}
+
+/*!
+  Compute the pose of a planar object from corresponding 2D-3D point coordinates and depth map.
+  Depth map is here used to estimate the 3D plane of the object.
+
+  \param[in] depthMap : Depth map aligned to the color image from where \e corners are extracted.
+  \param[in] corners : Vector of 2D pixel coordinates of the object in an image.
+  \param[in] colorIntrinsics : Camera parameters used to convert \e corners from pixel to meters.
+  \param[in] point3d : Vector of 3D points corresponding to the model of the planar object.
+  \param[out] cMo : Computed pose.
+  \param[out] confidence : Pose estimation confidence index in range [0, 1]. When values are close to zero
```

Comment:
I don't like the word `confidence`. It looks too simplistic to use the depth filling ratio as a confidence measure.
I don't have recent experience with the method. If you have introduced it because you encountered sometimes bad estimated pose with this method, I would propose to reword the parameter. I don't have a good proposition though (depth filling ratio?, depth data percentage?).

There are also the M-estimator weights that should give some information.

Reply:
Based on my experience, the "depth filling ratio" or better named the normalized number of depth data covering the tag is useful to detect when the depth sensor is occluded or partially occluded or when there are wholes in the depth map.

I agree that this is too simple to be a good confidence indicator.

A good compromise that I found is to compute the confidence index as the product between between the normalized number of depth data covering the tag and the normalized M-estimator weights returned by the robust estimation of the tag 3D plane. Results that I obtained sound good.